tests/pipeline/test_slice.py

```diff
@@ -0,0 +1,304 @@
+
+        class UsesSlicedInput(CustomFactor):
+            window_length = 1
+            inputs = [returns_slice]
+
+            def compute(self, today, assets, out, returns_slice):
+                pass
+
+        with self.assertRaises(NonExistentAssetInTimeFrame):
+            self.engine.run_pipeline(
+                Pipeline(columns={'uses_sliced_input': UsesSlicedInput()}),
+                self.start_date,
+                self.end_date,
+            )
+
+    def test_window_safety_of_slices(self):
+        """
+        Test that slices correctly inherit the `window_safe` property of the
+        term from which they are derived.
+        """
+        my_asset = self.asset_finder.retrieve_asset(self.sids[0])
+
+        # SimpleMovingAverage is not window safe.
+        sma = SimpleMovingAverage(
+            inputs=[USEquityPricing.close], window_length=10,
+        )
+        sma_slice = sma[my_asset]
+
+        class UsesSlicedInput(CustomFactor):
+            window_length = 1
+            inputs = [sma_slice]
+
+            def compute(self, today, assets, out, sma_slice):
+                pass
+
+        with self.assertRaises(NonWindowSafeInput):
+            self.engine.run_pipeline(
+                Pipeline(columns={'uses_sliced_input': UsesSlicedInput()}),
+                self.start_date,
+                self.end_date,
+            )
+
+        # Make sure that slices of custom factors are not window safe.
+        class MyFactor(CustomFactor):
+            window_length = 1
+            inputs = [USEquityPricing.close]
+
+            def compute(self, today, assets, out, close):
+                out[:] = close
+
+        my_factor = MyFactor()
+        my_factor_slice = my_factor[my_asset]
+
+        class UsesSlicedInput(CustomFactor):
+            window_length = 1
+            inputs = [my_factor_slice]
+
+            def compute(self, today, assets, out, my_factor_slice):
+                pass
+
+        with self.assertRaises(NonWindowSafeInput):
+            self.engine.run_pipeline(
+                Pipeline(columns={'uses_sliced_input': UsesSlicedInput()}),
+                self.start_date,
+                self.end_date,
+            )
+
+        # Slices of Returns *are* window safe.
+        returns = Returns(window_length=2)
+        returns_slice = returns[my_asset]
+
+        # Make sure that correlations are not safe if either the factor *or*
+        # the target slice are not window safe.
+        with self.assertRaises(NonWindowSafeInput):
+            my_factor.rolling_pearsonr(
+                target_slice=returns_slice,
+                correlation_length=10,
+            )
+
+        with self.assertRaises(NonWindowSafeInput):
+            returns.rolling_pearsonr(
+                target_slice=my_factor_slice,
+                correlation_length=10,
+            )
+
+    @parameter_space(returns_length=[2, 3], correlation_length=[3, 4])
+    def test_factor_correlation_methods(self,
```

Comment:
These tests aren't super representative since they're being tested against `arange` data, which means the correlation coefficients should be 1 for everything I think.  This whole suite might be a good use-case for `WithSeededRandomPipelineEngine`